### ISSUE TITLE
parent of things

### ISSUE THREAD

**Issue by marcoscaceres** [w3c/webidl2.js#85]
When processing IDL, it would be nice have a back reference to the parent/owner of things from the parse tree. 

Like, "operation" would have the parent "interface" as a reference. Enum values would have the Enum as the parent. Params would have the "operation" as the parent... and so on up the tree.

Interfaces, Enums, Dictionaries would just be `null` as the parent.

**Comment by marcoscaceres** [comment]
To be clear, I want to be able to do:

```
const interfaceObj = operation.parent;
if(interfaceObj.name === "Whatever"){...}; 
```

**Comment by SaschaNaz** [comment]
This will help implementing "[The [Replaceable] extended attribute must not be used on an attribute declared on a callback interface](https://heycam.github.io/webidl/#Replaceable)" restriction 👍

**Comment by marcoscaceres** [comment]
Played a bit with how to do this... but we might not be able to do this without significant changes (to tests): this would create a cyclic structure, breaking JSON output.

**Comment by SaschaNaz** [comment]
YAML is capable of cycle structure and jsondiffpatch also can cope with that. https://runkit.com/embed/mjk8jwzlgydl

I think transitioning to YAML is a viable choice, what do you think?

**Comment by marcoscaceres** [comment]
I'm ok to live without the parents (and happy to live with JSON for now). I wrote this algo to get around the limitation:

```JS
function addParent(obj, owner = null) {
  obj.parent = owner;
  Array.from(Object.entries(obj))
    .filter(([key]) => key !== "parent")
    .filter(([, value]) => Array.isArray(value) || value instanceof Object)
    .map(([, value]) => value)
    .forEach(value => {
      if (Array.isArray(value)) {
        value
          .filter(({ type }) => type !== "," && !whiteSpaceTypes.has(type))
          .forEach(item => {
            addParent(item, obj);
          });
        return;
      }
      value.parent = obj;
    });
}
```

**Comment by marcoscaceres** [comment]
Closing as won’t fix for now.

**Comment by marcoscaceres** [comment]
Continue to find this really useful in ReSpec and in the validator - having the parent gives really context ("The foo attribute of Interface Bar is missing X" and so one)... what I might do is exclude the parent when `JSON.stringify()` is called with the tests.

**Comment by SaschaNaz** [comment]
We can't test it then, an untested feature is prone to break. I think we ultimately should be able to test cyclic structures.

**Comment by SaschaNaz** [comment]
An option to remove (or add) cyclic structure would anyway be good to keep JSON compatibility.

**Comment by marcoscaceres** [comment]
Agree... we would need to have two modes.

**Comment by saschanaz** [comment]
I found this would be useful for https://github.com/w3c/respec/issues/2530 and also for simplifying writer code by skipping `parent` field.

**Comment by saschanaz** [comment]
Closed by #425
